<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="so_arm_101">

  <!-- Include materials -->
  <xacro:include filename="$(find so_arm_101_gazebo)/urdf/materials.xacro"/>

  <!-- Properties -->
  <xacro:property name="PI" value="3.14159265359"/>
  
  <!-- Joint properties based on original SO-ARM101 specs -->
  <xacro:property name="joint_effort" value="10.0"/>
  <xacro:property name="joint_velocity" value="10.0"/>

  <!-- Mesh scale -->
  <xacro:property name="mesh_scale" value="1.0 1.0 1.0"/>

  <!-- World Link (fixed to world) -->
  <link name="world"/>
  
  <!-- Base Link -->
  <link name="base">
    <inertial>
      <origin xyz="0.020739 0.00204287 0.065966" rpy="0 0 0"/>
      <mass value="0.147"/>
      <inertia ixx="0.0" ixy="0.0" ixz="0.0" 
               iyy="0.0" iyz="0.0" izz="0.0"/>
    </inertial>
    <!-- Base motor holder -->
    <visual>
      <origin xyz="0.0 0.0221255 0.0300817" rpy="1.5708 0.0 0.0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/base_motor_holder_so101_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_plastic"/>
    </visual>
    <!-- Base -->
    <visual>
      <origin xyz="0.0 0.0221255 0.0300817" rpy="1.5708 0 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/base_so101_v2.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_plastic"/>
    </visual>
    <!-- STS3215 Motor -->
    <visual>
      <origin xyz="0.0207909 -0.0105745 0.0761817" rpy="0.0 0.0 -1.5708"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/sts3215_03a_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_metal"/>
    </visual>
    <!-- Waveshare mounting plate -->
    <visual>
      <origin xyz="0.0 0.0467435 0.0798817" rpy="1.5708 0.0 0.0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/waveshare_mounting_plate_so101_v2.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_plastic"/>
    </visual>
    <collision>
      <origin xyz="0.0207909 0.0221255 0.0300817" rpy="1.5708 0 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/base_so101_v2.stl" scale="${mesh_scale}"/>
      </geometry>
    </collision>
  </link>

  <joint name="world_to_base" type="fixed">
    <parent link="world"/>
    <child link="base"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <!-- Joint 1: Shoulder Rotation -->
  <link name="shoulder">
    <inertial>
      <origin xyz="-0.0307604 0.0 -0.0252713" rpy="0 0 0"/>
      <mass value="0.100006"/>
      <inertia ixx="0.0" ixy="0.0" ixz="0.0" 
               iyy="0.0" iyz="0.0" izz="0.0"/>
    </inertial>
    <!-- STS3215 Motor -->
    <visual>
      <origin xyz="-0.0303992 0.0 -0.0417" rpy="1.5708 1.5708 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/sts3215_03a_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_metal"/>
    </visual>
    <!-- Motor holder base -->
    <visual>
      <origin xyz="-0.0675992 0.0 0.0158499" rpy="1.5708 -1.5708 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/motor_holder_so101_base_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_plastic"/>
    </visual>
    <!-- Rotation pitch -->
    <visual>
      <origin xyz="0.0122008 0.0 0.0464" rpy="-1.5708 0 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/rotation_pitch_so101_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_plastic"/>
    </visual>
    <collision>
      <origin xyz="0.0122008 0.0 0.0464" rpy="-1.5708 0 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/rotation_pitch_so101_v1.stl" scale="${mesh_scale}"/>
      </geometry>
    </collision>
  </link>

  <joint name="shoulder_rotation" type="revolute">
    <origin xyz="0.0 -0.0230745 0.0948817" rpy="-3.14159 0.0 1.5708"/>
    <parent link="base"/>
    <child link="shoulder"/>
    <axis xyz="0 0 1"/>
    <limit lower="-1.91986" upper="1.91986" effort="${joint_effort}" velocity="${joint_velocity}"/>
    <dynamics damping="0.6" friction="0.052"/>
  </joint>

  <!-- Joint 2: Shoulder Pitch -->
  <link name="upper_arm">
    <inertial>
      <origin xyz="-0.0898471 -0.00838224 0.0184089" rpy="0 0 0"/>
      <mass value="0.103"/>
      <inertia ixx="0.0" ixy="0.0" ixz="0.0" 
               iyy="0.0" iyz="0.0" izz="0.0"/>
    </inertial>
    <!-- STS3215 Motor -->
    <visual>
      <origin xyz="-0.11257 -0.0155 0.0187" rpy="-3.14159 0.0 -1.5708"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/sts3215_03a_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_metal"/>
    </visual>
    <!-- Upper arm -->
    <visual>
      <origin xyz="-0.065085 0.012 0.0182" rpy="3.14159 0.0 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/upper_arm_so101_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_plastic"/>
    </visual>
    <collision>
      <origin xyz="-0.065085 0.012 0.0182" rpy="3.14159 0.0 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/upper_arm_so101_v1.stl" scale="${mesh_scale}"/>
      </geometry>
    </collision>
  </link>

  <joint name="shoulder_pitch" type="revolute">
    <origin xyz="-0.0303992 -0.0182778 -0.0542" rpy="-1.5708 -1.5708 0"/>
    <parent link="shoulder"/>
    <child link="upper_arm"/>
    <axis xyz="0 0 1"/>
    <limit lower="-1.74533" upper="1.74533" effort="${joint_effort}" velocity="${joint_velocity}"/>
    <dynamics damping="0.6" friction="0.052"/>
  </joint>

  <!-- Joint 3: Elbow -->
  <link name="lower_arm">
    <inertial>
      <origin xyz="-0.0980701 0.00324376 0.0182831" rpy="0 0 0"/>
      <mass value="0.104"/>
      <inertia ixx="0.0" ixy="0.0" ixz="0.0" 
               iyy="0.0" iyz="0.0" izz="0.0"/>
    </inertial>
    <!-- Under arm -->
    <visual>
      <origin xyz="-0.0648499 -0.032 0.0182" rpy="-3.14159 0 0.0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/under_arm_so101_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_plastic"/>
    </visual>
    <!-- Motor holder wrist -->
    <visual>
      <origin xyz="-0.0648499 -0.032 0.018" rpy="-3.14159 0.0 0.0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/motor_holder_so101_wrist_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_plastic"/>
    </visual>
    <!-- STS3215 Motor -->
    <visual>
      <origin xyz="-0.1224 0.0052 0.0187" rpy="-3.14159 0.0 -3.14159"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/sts3215_03a_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_metal"/>
    </visual>
    <collision>
      <origin xyz="-0.0648499 -0.032 0.0182" rpy="-3.14159 0 0.0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/under_arm_so101_v1.stl" scale="${mesh_scale}"/>
      </geometry>
    </collision>
  </link>

  <joint name="elbow" type="revolute">
    <origin xyz="-0.11257 -0.028 0.0" rpy="0.0 0.0 1.5708"/>
    <parent link="upper_arm"/>
    <child link="lower_arm"/>
    <axis xyz="0 0 1"/>
    <limit lower="-1.74533" upper="1.74533" effort="${joint_effort}" velocity="${joint_velocity}"/>
    <dynamics damping="0.6" friction="0.052"/>
  </joint>

  <!-- Joint 4: Wrist Pitch -->
  <link name="wrist">
    <inertial>
      <origin xyz="0.0 -0.0386143 0.0281156" rpy="0 0 0"/>
      <mass value="0.079"/>
      <inertia ixx="0.0" ixy="0.0" ixz="0.0" 
               iyy="0.0" iyz="0.0" izz="0.0"/>
    </inertial>
    <!-- STS3215 Motor no horn -->
    <visual>
      <origin xyz="0.0 -0.0424 0.0306" rpy="1.5708 1.5708 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/sts3215_03a_no_horn_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_metal"/>
    </visual>
    <!-- Wrist roll pitch -->
    <visual>
      <origin xyz="0 -0.028 0.0181" rpy="-1.5708 -1.5708 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/wrist_roll_pitch_so101_v2.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_plastic"/>
    </visual>
    <collision>
      <origin xyz="0 -0.028 0.0181" rpy="-1.5708 -1.5708 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/wrist_roll_pitch_so101_v2.stl" scale="${mesh_scale}"/>
      </geometry>
    </collision>
  </link>

  <joint name="wrist_pitch" type="revolute">
    <origin xyz="-0.1349 0.0052 0.0" rpy="0.0 0.0 -1.5708"/>
    <parent link="lower_arm"/>
    <child link="wrist"/>
    <axis xyz="0 0 1"/>
    <limit lower="-1.65806" upper="1.65806" effort="${joint_effort}" velocity="${joint_velocity}"/>
    <dynamics damping="0.6" friction="0.052"/>
  </joint>

  <!-- Joint 5: Wrist Roll -->
  <link name="gripper">
    <inertial>
      <origin xyz="0.0 0.0 -0.025187" rpy="0 0 0"/>
      <mass value="0.087"/>
      <inertia ixx="0.0" ixy="0.0" ixz="0.0" 
               iyy="0.0" iyz="0.0" izz="0.0"/>
    </inertial>
    <!-- STS3215 Motor -->
    <visual>
      <origin xyz="0.0077 0.0 -0.0234" rpy="-1.5708 0.0 0.0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/sts3215_03a_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_metal"/>
    </visual>
    <!-- Wrist roll follower -->
    <visual>
      <origin xyz="0.0 0.0 0.0" rpy="-3.14159 0.0 0.0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/wrist_roll_follower_so101_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_plastic"/>
    </visual>
    <collision>
      <origin xyz="0.0 0.0 0.0" rpy="-3.14159 0.0 0.0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/wrist_roll_follower_so101_v1.stl" scale="${mesh_scale}"/>
      </geometry>
    </collision>
  </link>

  <joint name="wrist_roll" type="revolute">
    <origin xyz="0 -0.0611 0.0181" rpy="1.5708 0.0 3.14159"/>
    <parent link="wrist"/>
    <child link="gripper"/>
    <axis xyz="0 0 1"/>
    <limit lower="-2.79253" upper="2.79253" effort="${joint_effort}" velocity="${joint_velocity}"/>
    <dynamics damping="0.6" friction="0.052"/>
  </joint>

  <!-- Joint 6: Gripper -->
  <link name="gripper_jaw">
    <inertial>
      <origin xyz="-0.00157495 -0.0300244 0.0192755" rpy="0 0 0"/>
      <mass value="0.012"/>
      <inertia ixx="0.0" ixy="0.0" ixz="0.0" 
               iyy="0.0" iyz="0.0" izz="0.0"/>
    </inertial>
    <!-- Moving jaw -->
    <visual>
      <origin xyz="0.0 0.0 0.0189" rpy="0.0 0.0 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/moving_jaw_so101_v1.stl" scale="${mesh_scale}"/>
      </geometry>
      <material name="so101_plastic"/>
    </visual>
    <collision>
      <origin xyz="0.0 0.0 0.0189" rpy="0.0 0.0 0"/>
      <geometry>
        <mesh filename="package://so_arm_101_gazebo/meshes/moving_jaw_so101_v1.stl" scale="${mesh_scale}"/>
      </geometry>
    </collision>
  </link>

  <joint name="gripper_joint" type="revolute">
    <origin xyz="0.0202 0.0188 -0.0234" rpy="1.5708 0.0 0.0"/>
    <parent link="gripper"/>
    <child link="gripper_jaw"/>
    <axis xyz="0 0 1"/>
    <limit lower="-0.174533" upper="1.74533" effort="${joint_effort}" velocity="${joint_velocity}"/>
    <dynamics damping="0.6" friction="0.052"/>
  </joint>

</robot>
